### ISSUE TITLE
Question about the KL term in the loss function

### ISSUE THREAD

**Issue by rajat95** [ContextualAI/HALOs#4]
For y ~ p_chosen, the KL term for the loss is derived as  KL(p_policy(y_rejected|x') || p_reference(y_rejected|x'))), however according to technical report, the KL divergence should be calculated over the entire dataset. Is there a reason x' for above only includes only rejected inputs and not input x's from the entire batch (i.e include y_chosen too) ? I can guess maybe numerical stability could be the reason to make sure that two terms of the loss aren't correlated but want to make sure I am not missing something here.

**Comment by kawin-contextual-ai** [comment]
Great question!

In the current implementation, the KL term is calculated over each minibatch, not the whole dataset. For chosen example losses, the KL term is calculated with the rejected examples and vice-versa, as you point out. Because there's a 50/50 split of chosen:rejected examples in each minibatch (as enforced by the dataloader), this is possible. The reason we do this is that since we are also back-propagating through the KL term, the gradient wrt each example depends also on its contribution to the KL term. If we estimated the KL for each loss with all examples in the minibatch, then it would be possible for the gradient to increase the likelihood of a rejected example and decrease the likelihood of a chosen one. We don't ever want that, hence our implementation.

The big drawback of this is the need for a 50/50 split of chosen:rejected examples in each minibatch, which limits the scenarios in which KTO can be used.

To address this, tomorrow I will be pushing a better implementation: estimate the batch-wise (not minibatch-wise) KL by creating mismatched pairs of input-output pairs (x_i, y_{i+1}) and then do not backprop through this term. This gets rid of the current requirement for the 50:50 chosen:rejected split in each minibatch and the different KL estimate for chosen and rejected examples. It also works better in practice!

**Comment by kawin-contextual-ai** [comment]
@rajat95 The newer implementation I mentioned above has now been implemented (see [here](https://github.com/ContextualAI/HALOs/blob/44d82bbac74311a6cda0d934b20e7b860df89598/trainers.py#L777)).

i.e., We now estimate a single batchwise KL term for both the chosen and rejected losses, and we don't backprop through this KL term either. This means that we no longer need an even ratio of chosen: rejected examples in each minibatch, making KTO much easier to use. 

For example, we found that you can throw away about 90% of your chosen examples (while keeping all of your rejected examples) and still reach DPO-level performance with this new KTO implementation. Let me know if you have any more questions!
![image](https://github.com/ContextualAI/HALOs/assets/140556241/ec71ea17-46f9-4f8d-b242-4ffd4a13b0f4)

**Comment by rajat95** [comment]
thanks a lot for succint explanation!
